common/test/TestLogging.cs

```diff
@@ -2,39 +2,116 @@
+                    if (!string.IsNullOrWhiteSpace(iKey))
+                    {
+                        var config = new TelemetryConfiguration
+                        {
+                            InstrumentationKey = iKey,
+                        };
+                        _telemetryClient = new TelemetryClient(config);
+                        _testRunId = Guid.NewGuid().ToString();
+                    }
+                }
+            }
         }
 
         private const string NullInstance = "(null)";
-        
+
         public static string IdOf(object value) => value != null ? value.GetType().Name + "#" + GetHashCode(value) : NullInstance;
 
         public static int GetHashCode(object value) => value?.GetHashCode() ?? 0;
 
-        public static TestLogging GetInstance()
+        public static TestLogging GetInstance(TestContext testContext = null)
         {
-            return s_instance;
+            var logger = new TestLogging();
+
+            if (_telemetryClient != null)
+            {
+                // Set logger properties for the current test.
+                var targetFramework = (TargetFrameworkAttribute)Assembly
+                    .GetExecutingAssembly()
+                    .GetCustomAttributes(typeof(TargetFrameworkAttribute), false)
+                    .SingleOrDefault();
+                logger.Context = testContext;
+                logger.TargetFramework = targetFramework.FrameworkName;
+            };
+
+            return logger;
         }
 
-        public void WriteLine(string message)
+        public void WriteLine(string message, SeverityLevel severity = SeverityLevel.Information, IDictionary<string, string> propertyBag = null, [CallerMemberName] string caller = null)
         {
+            // Log to event source
             EventSourceTestLogging.Log.TestMessage(message);
+
+            // Log to Application insights
+            if (_telemetryClient != null)
+            {
+                if (propertyBag == null)
+                {
+                    propertyBag = new Dictionary<string, string>();
+                }
+
+                // Add common properties
+                propertyBag.Add(LoggingPropertyNames.TestRunId, _testRunId);
+                propertyBag.Add(LoggingPropertyNames.TestName, Context?.TestName);
+                propertyBag.Add(LoggingPropertyNames.ClassName, Context?.FullyQualifiedTestClassName ?? caller);
+                propertyBag.Add(LoggingPropertyNames.TestFramework, TargetFramework);
+                propertyBag.Add(LoggingPropertyNames.TestStatus, Context?.CurrentTestOutcome.ToString());
+                propertyBag.Add(LoggingPropertyNames.Language, Language);
+                propertyBag.Add(LoggingPropertyNames.Service, Service);
+
+                // Environment Variable set in Azure pipelines to correlate logs with the build number.
+                propertyBag.Add("BuildId", Environment.GetEnvironmentVariable("BUILD_BUILDID"));
```

Comment:
What if the variable is empty, because I ran locally?

Reply:
It will be null - No value. I thought that would be an indicator that this is not running in the pipeline. AI did not really crash with null values but i can add a check to be very safe.